airbyte-db/lib/src/main/java/io/airbyte/db/init/DatabaseInitializer.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.db.init;
+
+import static org.jooq.impl.DSL.select;
+
+import io.airbyte.db.Database;
+import io.airbyte.db.ExceptionWrappingDatabase;
+import io.airbyte.db.check.DatabaseAvailabilityCheck;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import org.jooq.DSLContext;
+import org.jooq.impl.DSL;
+import org.slf4j.Logger;
+
+/**
+ * Performs the initialization of the configured database if the database is available and has not
+ * yet been initialized.
+ *
+ * In the future, this logic could be completely removed if the schema initialization script is
+ * converted to a migration script.
+ */
+public interface DatabaseInitializer {
+
+  /**
+   * Initializes the configured database.
+   *
+   * @throws InterruptedException if unable to perform the initialization.
+   */
+  default void init() throws InterruptedException {
+    try {
+      initialize();
+    } catch (final IOException e) {
+      throw new InterruptedException("Unable to perform database initialization: " + e.getMessage());
+    }
+  }
+
+  /**
+   * Initializes the configured database by using the following steps:
+   *
+   * <ol>
+   * <li>Verify that the database is available and accepting connections</li>
+   * <li>Verify that the database is populated with the initial schema. If not, create the initial
+   * schema.</li>
+   * </ol>
+   *
+   * @throws IOException if unable to perform the initial schema check/creation.
+   * @throws InterruptedException if unable to verify the database availability.
+   */
+  default void initialize() throws IOException, InterruptedException {
+    // Verify that the database is up and reachable first
+    final Optional<DatabaseAvailabilityCheck> availabilityCheck = getDatabaseAvailabilityCheck();
+    if (availabilityCheck.isPresent()) {
+      availabilityCheck.get().check();
+      final Optional<DSLContext> dslContext = getDslContext();
+      if (dslContext.isPresent()) {
```

Comment:
Related to my above comment - it feels fairly messy to have all of these `isPresent()` checks, requiring us to nest the actual logic of the method multiple levels deep. Is there a path forward for removing the need for this?

Reply:
This is mainly a temporary fix, as we don't control the creation of these objects and they can therefore be `null`.  Once we convert to a dependency injection framework, the application will fail to start if a bean of the requested type/name cannot be found (unless of course we mark an injection point as optional).

The alternative in the current code is to put a bunch of ` != null` checks, which are just as messy and not the current, preferred way in Java to handle the potential for an object to not be present/set.  If we didn't care about the exceptions being raised, we would use `.ifPresent(o -> ...)` to remove the if checks.

Reply:
Got it, I think I had misunderstood [this comment](https://github.com/airbytehq/airbyte/pull/12961#discussion_r876201066) as saying that we would need to keep these checks even after we fully convert to dependency injection. I'm okay with this as a temporary fix